File: 0494-target-sum/0494-target-sum.cpp
class Solution {
public:
    
    int sol(int i,vector<int>& nums,int target2)
    {
       // if(target2==0)
       //     return 1;
        
        if(target2<0)
            return 0;
        
        if(i==0)
        {
            if(target2==0 && nums[i]==0)
                return 2;
            
            if(target2==0 || nums[i]==target2)
                return 1;
            
            else return 0;
                
            
        }
        
        // if(dp[i][target2]!=-999)return dp[i][target2];
        
        int notTake=sol(i-1,nums,target2);
        
        int take=0;
        //if(target2>=nums[i])
            take=sol(i-1,nums,target2-nums[i]);
        
        return take+notTake;
        
        
    }
    
    int findTargetSumWays(vector<int>& nums, int target) {
        
        int sum=0;
        for(int i=0;i<nums.size();i++)
            sum+=nums[i];
        
        if((sum+target)%2!=0)
            return 0;
        
        int target2=(sum+target)/2;
        
        // vector<vector<int>> dp(nums.size(),vector<int>(target2+1,-999));
        
        return sol(nums.size()-1,nums,target2);
        
    }
};

File: 0494-target-sum/0494-target-sum.java
// class Solution {
//     public int findTargetSumWays(int[] nums, int target) {
//         int sum=0;
//         for(int i : nums){
//             sum+=i;
//         }
//         int target2=(target+sum)/2;
//          if((sum+target)%2!=0)
//             return 0;
//         return backtrack(nums, target2, 0, new ArrayList<Integer>());
//     }
//     private static int backtrack(int[] nums, int target, int index, List<Integer> curr) {
//         if (index == nums.length) {
//             if (target == 0) {
//                 return 1;
//             }
//             return 0;
//         }
        
       
        
//         int pos=backtrack(nums, target - nums[index], index + 1, curr);
//         int neg=backtrack(nums, target, index + 1, curr);
       

//         return pos+neg;
//     }
// }
public class Solution {
    int count = 0;
    
    public int findTargetSumWays(int[] nums, int S) {
        calculate(nums, 0, 0, S);
        return count;
    }
    
    public void calculate(int[] nums, int i, int sum, int S) {
        if (i == nums.length) {
            if (sum == S) {
                count++;
            }
        } else {
            calculate(nums, i + 1, sum + nums[i], S);
            calculate(nums, i + 1, sum - nums[i], S);
        }
    }
}